optimism/src/mips/proof.rs

```diff
@@ -0,0 +1,444 @@
+use kimchi::{curve::KimchiCurve, groupmap::GroupMap};
+use mina_poseidon::sponge::ScalarChallenge;
+use mina_poseidon::FqSponge;
+use poly_commitment::{
+    commitment::{
+        absorb_commitment, combined_inner_product, BatchEvaluationProof, Evaluation, PolyComm,
+    },
+    evaluation_proof::DensePolynomialOrEvaluations,
+    OpenProof, SRS as _,
+};
+use rand::thread_rng;
+use rayon::iter::IntoParallelIterator;
+use rayon::iter::IntoParallelRefIterator;
+use rayon::iter::ParallelIterator;
+
+#[derive(Debug)]
+pub struct WitnessColumns<G> {
+    pub scratch: [G; crate::mips::witness::SCRATCH_SIZE],
+    pub instruction_counter: G,
+    pub error: G,
+}
+
+#[derive(Debug)]
+pub struct ProofInputs<G: KimchiCurve> {
+    evaluations: WitnessColumns<Vec<G::ScalarField>>,
+}
+
+impl<G: KimchiCurve> Default for ProofInputs<G> {
+    fn default() -> Self {
+        ProofInputs {
+            evaluations: WitnessColumns {
+                scratch: std::array::from_fn(|_| {
+                    (0..1 << 15).map(|_| G::ScalarField::zero()).collect()
+                }),
+                instruction_counter: (0..1 << 15).map(|_| G::ScalarField::zero()).collect(),
+                error: (0..1 << 15).map(|_| G::ScalarField::zero()).collect(),
+            },
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct Proof<G: KimchiCurve, OpeningProof: OpenProof<G>> {
+    commitments: WitnessColumns<PolyComm<G>>,
+    zeta_evaluations: WitnessColumns<G::ScalarField>,
+    zeta_omega_evaluations: WitnessColumns<G::ScalarField>,
+    opening_proof: OpeningProof,
+}
+
+pub fn fold<
+    G: KimchiCurve,
+    OpeningProof: OpenProof<G>,
+    EFqSponge: Clone + FqSponge<G::BaseField, G, G::ScalarField>,
+    EFrSponge: FrSponge<G::ScalarField>,
+>(
+    domain: EvaluationDomains<G::ScalarField>,
+    srs: &OpeningProof::SRS,
+    accumulator: &mut ProofInputs<G>,
+    inputs: WitnessColumns<Vec<G::ScalarField>>,
+) where
+    <OpeningProof as poly_commitment::OpenProof<G>>::SRS: std::marker::Sync,
+{
+    let polys = {
+        let WitnessColumns {
+            scratch,
+            instruction_counter,
+            error,
+        } = &inputs;
+        let eval_col = |evals: &Vec<G::ScalarField>| {
+            Evaluations::<G::ScalarField, D<G::ScalarField>>::from_vec_and_domain(
+                evals.clone(),
+                domain.d1,
+            )
+        };
+        let scratch = scratch.into_par_iter().map(eval_col).collect::<Vec<_>>();
+        WitnessColumns {
+            scratch: scratch.try_into().unwrap(),
+            instruction_counter: eval_col(instruction_counter),
+            error: eval_col(error),
+        }
+    };
+    let commitments = {
+        let WitnessColumns {
+            scratch,
+            instruction_counter,
+            error,
+        } = &polys;
+        let comm = |evals: &Evaluations<G::ScalarField, D<G::ScalarField>>| {
+            srs.commit_evaluations_non_hiding(domain.d1, evals)
```

Comment:
non hiding because Optimism is not asking for zk?

Reply:
Non-hiding because it simplifies things for now.